oshi-core-java25/src/main/java/oshi/ffm/driver/windows/registry/InstalledAppsDataFFM.java

```diff
@@ -0,0 +1,221 @@
+                                String installLocation = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "InstallLocation", accessFlag));
+                                String installSource = registryValueToString(
+                                        getRegistryValueOrNull(rootKey, fullPath, "InstallSource", accessFlag));
+
+                                Map<String, String> additionalInfo = new LinkedHashMap<>();
+                                additionalInfo.put("installLocation", installLocation);
+                                additionalInfo.put("installSource", installSource);
+
+                                ApplicationInfo app = new ApplicationInfo(name, version, publisher, installDate,
+                                        additionalInfo);
+                                appInfoSet.add(app);
+
+                            } catch (Throwable e) {
+                                LOG.trace("Skipping key {}: {}", fullPath, e.getMessage());
+                            }
+                        }
+                    } catch (Throwable e) {
+                        LOG.trace("Skipping path {}: {}", registryPath, e.getMessage());
+                    }
+                }
+            }
+        }
+
+        return new ArrayList<>(appInfoSet);
+    }
+
+    private static String registryValueToString(Object registryValueOrNull) {
+        if (registryValueOrNull instanceof Integer i) {
+            return Integer.toString(i);
+        }
+        return (String) registryValueOrNull;
+    }
+
+    private static long registryValueToLong(Object registryValueOrNull) {
+        if (registryValueOrNull == null) {
+            return 0L;
+        }
+        long currentTimeSecs = System.currentTimeMillis() / 1000L;
+        long minSaneTimestamp = currentTimeSecs - THIRTY_YEARS_IN_SECS;
+
+        if (registryValueOrNull instanceof Integer i) {
+            if (i > minSaneTimestamp && i < currentTimeSecs) {
+                return i * 1000L;
+            }
+            return i;
+        } else if (registryValueOrNull instanceof String s) {
+            String dateStr = s.trim();
+            long epoch = ParseUtil.parseDateToEpoch(dateStr, "yyyyMMdd");
+            if (epoch == 0) {
+                epoch = ParseUtil.parseDateToEpoch(dateStr, "MM/dd/yyyy");
+            }
+            return epoch;
+        }
+        return 0L;
+    }
+
+    private static Object getRegistryValueOrNull(MemorySegment rootKey, String path, String key, int accessFlag)
+            throws Throwable {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment phkKey = arena.allocate(ADDRESS);
+            int rc = RegOpenKeyEx(rootKey, toWideString(arena, path), 0, KEY_READ | accessFlag, phkKey);
+            if (rc != ERROR_SUCCESS) {
+                throw new Win32Exception(rc);
+            }
```

Comment:
Seems to me we could simply this pattern with a static method, possibly in a util class.
```suggestion
            checkSuccess(RegOpenKeyEx(rootKey, toWideString(arena, path), 0, KEY_READ | accessFlag, phkKey));
```
where somewhere we have a method
```java
public static void checkSuccess(int rc) {
    if (rc != ERROR_SUCCESS) {
        throw new Win32Exception(rc);    
    }
}
```